go/adbc/driver/bigquery/connection.go

```diff
@@ -0,0 +1,708 @@
+		}
+		structType := arrow.StructOf(nestedFields...)
+		if structType == nil {
+			return arrow.Field{}, adbc.Error{
+				Code: adbc.StatusInvalidArgument,
+				Msg:  fmt.Sprintf("Cannot create a struct schema for record `%s`", fieldRecords),
+			}
+		}
+		field.Type = structType
+	default:
+		return arrow.Field{}, adbc.Error{
+			Code: adbc.StatusInvalidArgument,
+			Msg:  fmt.Sprintf("Cannot handle data type `%s`, type=`%s`", dataType, typeString),
+		}
+	}
+	return field, nil
+}
+
+func parsePrecisionAndScale(name, typeString string) (int32, int32, error) {
+	typeString = strings.TrimSpace(typeString)
+	if len(typeString) == 0 {
+		return 0, 0, adbc.Error{
+			Code: adbc.StatusInvalidData,
+			Msg:  fmt.Sprintf("Cannot parse data type `%s` for field `%s`", typeString, name),
+		}
+	}
+
+	if typeString == "NUMERIC" {
+		return 38, 9, nil
+	} else if typeString == "BIGNUMERIC" {
+		return 76, 38, nil
+	}
+
+	r := precisionScaleRegex.FindStringSubmatch(typeString)
+	if len(r) != 3 {
+		return 0, 0, adbc.Error{
+			Code: adbc.StatusInvalidData,
+			Msg:  fmt.Sprintf("Cannot parse data type `%s` for field `%s`", typeString, name),
+		}
+	}
+
+	precisionString := r[precisionScaleRegex.SubexpIndex("precision")]
+	precision, err := strconv.ParseInt(precisionString, 10, 32)
+	if err != nil {
+		return 0, 0, adbc.Error{
+			Code: adbc.StatusInvalidData,
+			Msg:  fmt.Sprintf("Cannot parse precision `%s` for field `%s`: %s", precisionString, name, err.Error()),
+		}
+	}
+
+	scaleString := r[precisionScaleRegex.SubexpIndex("scale")]
+	scale, err := strconv.ParseInt(scaleString, 10, 32)
+	if err != nil {
+		return 0, 0, adbc.Error{
+			Code: adbc.StatusInvalidData,
+			Msg:  fmt.Sprintf("Cannot parse scale `%s` for field `%s`: %s", scaleString, name, err.Error()),
+		}
+	}
+	return int32(precision), int32(scale), nil
+}
+
+func (c *connectionImpl) getTableSchemaWithFilter(ctx context.Context, catalog *string, dbSchema *string, tableName string, columnName *string) (*arrow.Schema, error) {
+	if catalog == nil {
+		catalog = &c.catalog
+	}
+
+	if dbSchema == nil {
+		dbSchema = &c.dbSchema
+	}
+	sanitizedDbSchema, err := sanitizeDataset(*dbSchema)
+	if err != nil {
+		return nil, err
+	}
+
+	// sadly that query parameters cannot be used in place of table names
+	queryString := fmt.Sprintf("SELECT * FROM `%s`.`%s`.INFORMATION_SCHEMA.COLUMNS WHERE table_name = @tableName", *catalog, sanitizedDbSchema)
+	query := c.client.Query(queryString)
+	query.Parameters = []bigquery.QueryParameter{
+		{
+			Name:  "tableName",
+			Value: tableName,
+		},
+	}
+	reader, _, err := newRecordReader(ctx, query, c.Alloc)
+	if err != nil {
+		return nil, err
+	}
+
+	fields := make([]arrow.Field, 0)
+	var columns map[string]int = nil
+
+	for reader.Next() && ctx.Err() == nil {
+		rec := reader.Record()
+		if columns == nil {
+			columns = make(map[string]int)
+			for i, f := range reader.Schema().Fields() {
+				columns[strings.ToUpper(f.Name)] = i
+			}
+		}
```

Comment:
the schema already has functionality for retrieving fields by name (it internally maintains a map index) so you shouldn't need to manually construct this mapping.

If the issue is that the casing is unknown, then at worst you could construct this mapping outside of the loop as record readers have a `Schema` method